DOC: Correct selected C docstrings to eliminate warnings

eagunn opened the pull request:
Addresses #27377, eliminating all C docstring warning messages.

This is a minimal set of changes, simply eliminating the observed warning messages, with no attempt to improve C docstring content overall, per advice of @mattip.

Reviewers, 
- Some changes were straightforward:
-- The name of the `@param` variable to be documented was missing. I inserted it.
-- The name of the `@param` variable was misspelled. I corrected it.
-- An `@param` line documented a variable that no longer existed or had been split into two. I made the obvious fix.
-- In one case, I removed a docstring from an .h file that simply duplicated the docstring in the corresponding .c file.
- The changes to pay more attention to are those where I had to supply missing Description text. For those, I simply took my best shot and used as few words as I could, figuring the less I wrote, the fewer things would be said incorrectly.

To verify the error messages were being eliminated, I used the same commands as were documented in the original issue but piped the sorted, unique warnings to a file which I diffed against the original list as I worked.
$ CC=clang CXX=clang++ CFLAGS=-Wdocumentation pip install --log PIP.log .
$ grep -- -Wdocumentation PIP.log | sed -E -e 's/^\S+\s+//' | sort -n | uniq > warnings.list

[skip actions] [skip azp] [skip cirrus]

eagunn committed:
DOC: Correct selected C docstrings to eliminate warnings

Addresses #27377

seberg commented on numpy/_core/src/multiarray/descriptor.c:
```diff
@@ -1441,9 +1442,9 @@ descr_is_legacy_parametric_instance(PyArray_Descr *descr,
- * @param out_DType
+ * @param dtype Input descriptor to be converted
+ * @param out_descr Output descriptor
+ * @param out_DType Dtype of the output descriptor
```
```suggestion
 * @param out_DType DType of the output descriptor
```

seberg commented on numpy/_core/src/multiarray/descriptor.c:
```diff
@@ -1409,7 +1409,8 @@ PyArray_DescrConverter2(PyObject *obj, PyArray_Descr **at)
  *
- * @param descr
+ * @param descr descriptor to be checked
+ * @param DType pointer to the Dtype of the descriptor
```
```suggestion
 * @param DType pointer to the DType of the descriptor
```
I would prefer to try to use the `DType` capitalization consistently also in flowing text.  (in part because that is what composite words in german would do maybe, but mainly, I think it is clearer to use the code-name where even before me, `DataType` is far more common than `Datatype` in NumPy.)

seberg commented on numpy/_core/src/multiarray/convert_datatype.c:
```diff
@@ -793,11 +794,11 @@ npy_casting_to_string(NPY_CASTING casting)
- * @param scalar Whether this was a "scalar" cast (includes 0-D array with
- *               PyArray_CanCastArrayTo result).
+ * @param src_dtype The source descriptor to cast from
+ * @param dst_dtype The destination type trying to be cast to
```
```suggestion
 * @param dst_dtype The destination descriptor trying to cast to
```
Let's stick with "descriptor" to be precise if we do it before.  (the be seemed off to me, but not quite sure)

seberg commented on numpy/_core/src/multiarray/convert_datatype.c:
```diff
@@ -793,11 +794,11 @@ npy_casting_to_string(NPY_CASTING casting)
+ * @param src_dtype The source descriptor to cast from
+ * @param dst_dtype The destination type trying to be cast to
+ * @param casting The casting rule that was violated
+ * @param scalar Boolean flag indicating if this was a "scalar" cast (includes 0-D 
+ *               array withPyArray_CanCastArrayTo result).
```
```suggestion
 * @param scalar Boolean flag indicating if this was a "scalar" cast.
```
The bracket just got outdated with NumPy 2.

seberg commented on numpy/_core/src/multiarray/mapping.c:
```diff
@@ -814,11 +814,12 @@ get_item_pointer(PyArrayObject *self, char **ptr,
+ * @param view Resulting array (new reference)
+ * @param indices parsed index information
+ * @param index_num number of indices
+ * @param ensure_array true if result should have type PArray, 
```
```suggestion
 * @param ensure_array true if result should be a base class array, 
```

seberg commented on numpy/_core/src/multiarray/usertypes.c:
```diff
@@ -618,8 +618,8 @@ legacy_userdtype_common_dtype_function(
  *
- * @param from
- * @param to
+ * @param from Source type
+ * @param to Destination type
```
```suggestion
 * @param from Source DType
 * @param to Destination DType
```
Let's use DType/descriptor to be precise were possible.

seberg commented on numpy/_core/src/multiarray/convert_datatype.c:
```diff
@@ -409,11 +409,12 @@ _get_cast_safety_from_castingimpl(PyArrayMethodObject *castingimpl,
  * @param to_dtype If `to` is NULL, must pass the to_dtype (otherwise this
  *        is ignored).
- * @param[out] view_offset
+ * @param view_offset The byte offset, if any, between the original 
+ *        view/type and the view/type that would be cast to.
```
```suggestion
 * @param view_offset If set, the cast can be described by a view with
 *        this byte offset.  For example, casting "i8" to "i8,"
 *        (the structured dtype) can be described with `*view_offset = 0`.
```
Maybe to make it a bit clearer, since this is a tricky concept.

seberg commented on numpy/_core/src/multiarray/common_dtype.c:
```diff
@@ -106,7 +106,7 @@ PyArray_CommonDType(PyArray_DTypeMeta *dtype1, PyArray_DTypeMeta *dtype2)
  *
  * @param length Number of DTypes
- * @param dtypes
+ * @param dtypes List of Dtypes to be reduced
```
```suggestion
 * @param dtypes List of DTypes to be reduced
```

seberg reviewed (commented):
Thank you, looks great!  I added a few comments/suggestions, you don't need to adopt them all.

I do much prefer the "DType" spelling, and I think we are using it far more often right now, otherwise some smaller suggestions or just trying to expand/clarify a bit.

eagunn committed:
Apply suggestions from @seberg code review

Many are simple corrections for the spelling for `DType`
A few are changes of substance to clarify a Definition.

Co-authored-by: Sebastian Berg <sebastian@sipsolutions.net>

eagunn commented:
@seberg Thank you for the detailed review. I read through all your suggestions carefully, then accepted them all.
Thanks for the clarification on correct capitalization for DType. I was wondering if all the variations bothered others as they do me but went for minimal changes as I worked. I'm much happier to see a layer of consistency be added.

eagunn committed:
Merge branch 'numpy:main' into c-doc-warnings

seberg commented:
Thanks a lot @eagunn, let's put it in :).

eagunn commented:
Huzzah. Thank you.
